Add group_row_collapse function

thebioengineer opened the pull request:
Add a function that will take a data.frame and column names and will collapse the duplicated rows into "empties" for pretty table printing. 

I included the associated tests/docs that will resolve #44

monicagerber commented:
Hi @thebioengineer. Will take a look and review tomorrow.

thebioengineer commented:
Thanks @monicagerber! Hoping this function will be able to assist people

monicagerber commented on R/group_row_collapse.R:
```diff
@@ -0,0 +1,51 @@
+#' Output row-collapsed tables for printing
+#'
+#' escape takes a vector of chracter values, and puts "\\" in front of them to make them
```
I think you copied this from elsewhere and forgot to delete.

thebioengineer replied:
Thanks for pointing that out - thought I'd updated that!

monicagerber commented on R/group_row_collapse.R:
```diff
@@ -0,0 +1,51 @@
+#' Output row-collapsed tables for printing
```
The description is a bit vague. How about something along the lines of:
"Replace repeated rows in a data.frame with NA."

thebioengineer replied:
> * Can you make sure that your code follows the style guide? For example, [spacing](https://style.tidyverse.org/syntax.html#spacing) and formatting of [long lines](https://style.tidyverse.org/syntax.html#long-lines) needs to be fixed.

Fixed - forgot to run formatR

> * The function is a little hard for me to follow. It would be helpful for me to understand why you are doing things in a couple of places. For example, line 32 (sorting row order) and then creating new fields and dropping old fields. Why is it necessary to create new fields?

This was an artifact when I as originally writing this using tidyverse (group_by and mutate) - I rewrote it in base and spaced that the limitation of editing one of the grouping fields is only limited in tidyverse.

> * Making the solution more generalizable to other classes.
> * Is function name ok? In VISCfunctions there is more of a verb + object pattern (`collapse_group_row`?)

Updated to `collapse_group_row`, I see the argument for that. As for making it more generalizable, are you suggesting for data.table? We could use methods, so we have a data.frame and data.table method?

monicagerber commented on R/group_row_collapse.R:
```diff
@@ -0,0 +1,51 @@
+#' @param .data a data.frame
+#' @param ... Variables to collapse rows by
+#'
+#' @return a data.frame where duplicated rows identified by ... are replaced with ""
```
The function I have for this replaces duplicated rows with NA's. Will your function change the class of the column? What happens if the column is a factor variable? I think we may need to make this more robust to other classes. We could discuss during Thursday stand up.

thebioengineer commented on R/group_row_collapse.R:
```diff
@@ -0,0 +1,51 @@
+  stopifnot(is.data.frame(.data))
+
+  fields<-as.character(as.list(substitute(substitute(...)))[-1])
+  fields<-gsub("`","",fields)
```
I added comments explaining this. Basically it allows for us to use NSE ->

collapse_group_row(data,x,y,z) vs collapse_group_row(data,"x","y","z")

thebioengineer commented:
> * Can you make sure that your code follows the style guide? For example, [spacing](https://style.tidyverse.org/syntax.html#spacing) and formatting of [long lines](https://style.tidyverse.org/syntax.html#long-lines) needs to be fixed.

Fixed - forgot to run formatR

> * The function is a little hard for me to follow. It would be helpful for me to understand why you are doing things in a couple of places. For example, line 32 (sorting row order) and then creating new fields and dropping old fields. Why is it necessary to create new fields?

This was an artifact when I as originally writing this using tidyverse (group_by and mutate) - I rewrote it in base and spaced that the limitation of editing one of the grouping fields is only limited in tidyverse.

> * Making the solution more generalizable to other classes.
> * Is function name ok? In VISCfunctions there is more of a verb + object pattern (`collapse_group_row`?)

Updated to `collapse_group_row`, I see the argument for that. As for making it more generalizable, are you suggesting for data.table? We could use methods, so we have a data.frame and data.table method?

monicagerber commented:
Hi @thebioengineer, 

re: your last question, also see my comment above on line 9. I am suggesting that it needs to work with different column classes (e.g., factor and numeric).

thebioengineer commented:
@monicagerber I see. 

I changed the function so that the class of the columns it is grouping by does not change, it simply drops NA's into the replicated rows. So the functionl order them based on how R would naturally order a factor and numeric using the `order()` function.

thebioengineer commented on R/collapse_group_row.R:
```diff
@@ -0,0 +1,49 @@
+#'   outputVal = runif(7)
+#' )
+#'
+#' group_row_collapse(sample_df, x, y, z)
```
Sounds good.

monicagerber commented:
Thanks @wfulp for taking a look too. All the improvements look great, @thebioengineer. 

One thing that is different about this function compared to `kableExtra::collapse_rows()` is that it will re-order the columns if you pick only `y` to collapse:

```
sample_df <- data.frame(
       x = c(1, 1, 1, 2, 2, 2, 2),
       y = c('test1', 'test1', 'test2', 'test1', 'test2', 'test2', 'test1'),
       z = c(1, 2, 3, 4, 5, 6, 7),
       outputVal = runif(7)
     )
collapse_group_row(sample_df, y)

kable(sample_df) %>% collapse_rows(2)
```
Maybe a minor inconvenience to have to re-order the columns. I'll let @wfulp decided if he's ok with that.

wfulp commented:
> Thanks @wfulp for taking a look too. All the improvements look great, @thebioengineer.
> 
> One thing that is different about this function compared to `kableExtra::collapse_rows()` is that it will re-order the columns if you pick only `y` to collapse:
> 
> ```
> sample_df <- data.frame(
>        x = c(1, 1, 1, 2, 2, 2, 2),
>        y = c('test1', 'test1', 'test2', 'test1', 'test2', 'test2', 'test1'),
>        z = c(1, 2, 3, 4, 5, 6, 7),
>        outputVal = runif(7)
>      )
> collapse_group_row(sample_df, y)
> 
> kable(sample_df) %>% collapse_rows(2)
> ```
> 
> Maybe a minor inconvenience to have to re-order the columns. I'll let @wfulp decided if he's ok with that.

Good catch @monicagerber . It does seems strange the function forces the collapsed fields to the front. Maybe @thebioengineer uses this to both sort columns and collapse in one call. This does make it a pain if you ever want to collapse a field in the middle of the dataset. I'll let @thebioengineer make the call if he actually uses this functions to bring fields to the front. If not I say we keep it consistent with `kableExtra::collapse_rows()`

Note if we do decide to make the change it just requires replacing the last line of the function:
```.data[, c(paste0(fields), alt_fields)]```
 with: 
```.data```

thebioengineer commented:
I have not had any cases where I would want to collapse and not move them over to the "front" of the data.frame, which is why I wrote the function in that way. Not sure of cases where you would want to collapse a middle row and not have it be on the outside? In the example shown, I would want to collapse by X and Y.

wfulp commented:
As a compromise I added a `reorder_col = TRUE` param. @monicagerber @thebioengineer any objections?

thebioengineer commented:
Sounds good to me

monicagerber commented:
I think that is a good compromise, @wfulp. Looks like you already merged in though. ;)
